src/chapters/control-flow-and-collections/arraylist.rst

```diff
@@ -0,0 +1,371 @@
+
+.. index:: ! generic class, generic type
+
+.. note:: 
+
+   You will sometimes see the ``ArrayList`` class written as ArrayList<E>,
+   where ``E`` represents a placeholder for the type that a programmer will
+   declare a given list to hold. This is especially true in documentation.
+   You can think of ``E`` as representing an arbitrary type.
+
+   Classes like ArrayList<E> that take another type or class as a parameter
+   are referred to as **generic classes** or **generic types**.
+
+
+We then use a ``do-while`` loop to collect the names of each of the students
+in the class.
+
+.. sourcecode:: java
+   :lineno-start: 18
+
+   // Get student names
+   do {
+      newStudent = in.nextLine();
+
+      if (!newStudent.equals("")) {
+         students.add(newStudent);
+      }
+
+   } while(!newStudent.equals(""));
+
+Recall that a ``do-while`` loop is very similar to a ``while`` loop, but 
+the execution condition is checked at the end of the loop block. This has 
+the net effect that the code block will always run at least once. In this example, 
+we prompt the user for a name, which Java processes via ``in.nextLine()`` when the 
+user hits the enter key. To finish entering names, the user enters a blank line.
+
+.. index:: ! ArrayList.add()
+
+For each student that is entered (that is, each non-empty line), we add
+the new String to the end of our list with ``students.add(newStudent)``.
+The ``.add()`` method is provided by the `ArrayList
+Class <http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html>`__.
+There are lots of other ArrayList methods to get familiar with, some of which we will discuss
+in more detail below.
+
+Note that our program imports ``java.util.ArrayList`` to take advantage of this Java
+provided class.
+
+Below the ``do-while`` loop are two different loops that demonstrate two ways you 
+can loop through a list in Java. Here’s the first, which
+collects the numeric grade for each student:
+
+.. sourcecode:: java
+   :lineno-start: 27
+
+   // Get student grades
+   for (String student : students) {
+      System.out.print("Grade for " + student + ": ");
+      Double grade = in.nextDouble();
+      grades.add(grade);
+   }
+
+This, you may recall, is Java's ``for-each`` loop syntax. You may read this 
+in your head, or even aloud, as: ``for each student in students``. As you might
+expect at this point, we must declare the iterator variable ``student``
+with its data type.
+
+The next loop on display prints out each student’s name and grade:
+
+.. sourcecode:: java
+   :lineno-start: 34
+
+   // Print class roster
+   System.out.println("\nClass roster:");
+   double sum = 0.0;
+
+   for (int i = 0; i < students.size(); i++) {
+      System.out.println(students.get(i) + " (" + grades.get(i) + ")");
+      sum += grades.get(i);
+   }
+
+.. index:: ! ArrayList.size()
+
+Here, we introduce the syntax ``students.size()`` which utilizes the
+``ArrayList``\ ’s ``size()`` method to return the integer representing
```

Comment:
I'm not sure how much I like the mixing of highlight, 's, highlight...

Actually, I am sure.  Can we avoid this?